server/src/main/java/org/elasticsearch/health/HealthPeriodicLogger.java

```diff
@@ -0,0 +1,194 @@
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.gateway.GatewayService;
+import org.elasticsearch.health.node.selection.HealthNode;
+
+import java.io.Closeable;
+import java.time.Clock;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+public class HealthPeriodicLogger implements ClusterStateListener, Closeable, SchedulerEngine.Listener {
+    public static final String HEALTH_SERVICE_POLL_INTERVAL = "health_service.poll_interval";
+    public static final Setting<TimeValue> HEALTH_SERVICE_POLL_INTERVAL_SETTING = Setting.timeSetting(
+        HEALTH_SERVICE_POLL_INTERVAL,
+        TimeValue.timeValueSeconds(60),
+        TimeValue.timeValueSeconds(15),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Name constant for the job HealthService schedules
+     */
+    private static final String HEALTH_SERVICE_JOB_NAME = "health_service";
+
+    private final Settings settings;
+
+    private final ClusterService clusterService;
+    private final NodeClient client;
+
+    private final HealthService healthService;
+    private final Clock clock;
+
+    private volatile boolean isHealthNode = false;
+    private SchedulerEngine.Job scheduledJob;
+    private final SetOnce<SchedulerEngine> scheduler = new SetOnce<>();
+    private volatile TimeValue pollInterval;
+
+    private static final Logger logger = LogManager.getLogger(HealthPeriodicLogger.class);
+
+    public HealthPeriodicLogger(Settings settings, ClusterService clusterService, NodeClient client, HealthService healthService) {
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.client = client;
+        this.healthService = healthService;
+        this.clock = getClock();
+        this.scheduledJob = null;
+        this.pollInterval = HEALTH_SERVICE_POLL_INTERVAL_SETTING.get(settings);
+    }
+
+    /**
+     * Initializer method to avoid the publication of a self reference in the constructor.
+     */
+    public void init() {
+        clusterService.addListener(this);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(HEALTH_SERVICE_POLL_INTERVAL_SETTING, this::updatePollInterval);
+    }
+
+    private void updatePollInterval(TimeValue newInterval) {
+        this.pollInterval = newInterval;
+        maybeScheduleJob();
+    }
+
+    protected Clock getClock() {
+        return Clock.systemUTC();
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // wait for the cluster state to be recovered
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
+        DiscoveryNode healthNode = HealthNode.findHealthNode(event.state());
+        if (healthNode == null) {
+            return;
+        }
```

Comment:
This should probably cancel the job, if no health node is found. Otherwise if the task was previously scheduled on this node since it's the health node, and then that health task gets unallocated (so the health node is `null`), the logging task will never be cancelled.